processor/samplingprocessor/tailsamplingprocessor/processor.go

```diff
@@ -244,37 +241,47 @@ func (tsp *tailSamplingSpanProcessor) makeDecision(id pdata.TraceID, trace *samp
 }
 
 // ConsumeTraceData is required by the SpanProcessor interface.
-func (tsp *tailSamplingSpanProcessor) ConsumeTraces(_ context.Context, td pdata.Traces) error {
+func (tsp *tailSamplingSpanProcessor) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
 	tsp.start.Do(func() {
 		tsp.logger.Info("First trace data arrived, starting tail_sampling timers")
 		tsp.policyTicker.Start(1 * time.Second)
 	})
-
-	octds := internaldata.TraceDataToOC(td)
-	var errs []error
-	for _, octd := range octds {
-		if err := tsp.processTraces(octd); err != nil {
-			errs = append(errs, err)
+	resourceSpans := td.ResourceSpans()
+	for i := 0; i < resourceSpans.Len(); i++ {
+		resourceSpan := resourceSpans.At(i)
+		if resourceSpan.IsNil() {
+			continue
 		}
+		tsp.processTraces(resourceSpan)
 	}
-	return componenterror.CombineErrors(errs)
+	return nil
 }
 
-func (tsp *tailSamplingSpanProcessor) processTraces(td consumerdata.TraceData) error {
-
-	// Groupd spans per their traceId to minimize contention on idToTrace
-	idToSpans := make(map[traceKey][]*tracepb.Span)
-	for _, span := range td.Spans {
-		if len(span.TraceId) != 16 {
-			tsp.logger.Warn("Span without valid TraceId", zap.String("SourceFormat", td.SourceFormat))
+func (tsp *tailSamplingSpanProcessor) groupSpansByTraceKey(resourceSpans pdata.ResourceSpans) map[traceKey][]*pdata.Span {
```

Comment:
Note: we would probably remove this logic in a next iteration of the processor, especially once https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/1235 is done. For instance, the instrumentation library name isn't being kept by this logic here, and I think it should.

Reply:
Sure